pre-check/v1.1.x/check.sh

```diff
@@ -0,0 +1,310 @@
+#!/bin/bash -e
+
+check_failed=0
+
+record_fail()
+{
+    check_failed=$((check_failed+1))
+}
+
+
+check_bundles()
+{
+    echo ">>> Check all bundles ready..."
+
+    # fleet bundles should be ready
+    # except mcc-harvester, which has Modified status in 1.0.3
+    bundles=$(kubectl get bundles.fleet.cattle.io -A -o yaml)
+    pending=$(echo "$bundles" | yq '.items | any_c(.metadata.name != "mcc-harvester" and .spec.helm != null and .status.summary.ready == 0)')
+
+    if [ "$pending" = "false" ]; then
+      echo "All Helm bundles are ready."
+      return
+    fi
+
+    echo "There are non-ready Helm bundles:"
+    echo "$bundles" | yq '.items[] | select(.spec.helm != null and .status.summary.ready == 0) | .metadata.namespace + "/" + .metadata.name'
+    record_fail
+}
+
+check_harvester_bundle()
+{
+    echo ">>> Check the Harvester bundle is ready..."
+
+    current_summary=$(mktemp)
+    kubectl get bundles.fleet.cattle.io/mcc-harvester -n fleet-local -o yaml | yq '.status.summary' > $current_summary
+
+    expected_summary=$(mktemp)
+cat > $expected_summary <<EOF
+desiredReady: 1
+ready: 1
+EOF
+
+    if ! diff <(yq -P 'sort_keys(..)' $current_summary) <(yq -P 'sort_keys(..)' $expected_summary); then
+        echo "Harvester bundle is not ready!"
+        cat $current_summary
+        record_fail
+        return
+    fi
+
+    echo "The Harvester bundle is ready."
+}
+
+check_nodes()
+{
+    local failed="false"
+    echo ">>> Check all nodes are ready..."
+
+    # Use a file to store the node state becuase we can't set the global variable inside the piped scope
+    # The file is removed in the piped scope if there are not-ready nodes.
+    node_ready_state=$(mktemp)
+
+    # nodes should not be cordoned
+    nodes=$(kubectl get nodes -o yaml)
+    unschedulable=$(echo "$nodes" | yq '.items | any_c(.spec.unschedulable == true)')
+
+    if [ "$unschedulable" = "true" ]; then
+        echo "There are unschedulable nodes:"
+        echo "$nodes" | yq '.items[] | select(.spec.unschedulable == true)  | .metadata.name'
+        rm -f $node_ready_state
+    fi
```

Comment:
Besides `.spec.unschedulable`, do we want also to check the KubeVirt taint?

Reply:
I guess we can check if there is any taint. A tainted node might cause unexpected behavior.

Reply:
Added in https://github.com/harvester/upgrade-helpers/pull/5/commits/42e9352541b8a2f8f4398b5a69d648289b5f076e